### PR TITLE
Provide options to allow for customising the iOS/macOS specific session settings

### LINKED ISSUE #53: iPhone 12 playing back on earpiece speaker and therefore is very quiet.
When running on the iPhone 12 the playback volume is very low. 

```
        _audioPlayer = AudioManager.Current.CreatePlayer(filePath);
        _audioPlayer.Play();
```

Even if I try manually setting the volume to 1 it is still the same low volume.

```
        _audioPlayer = AudioManager.Current.CreatePlayer(filePath);
        _audioPlayer.Volume = 1;
        _audioPlayer.Play();
```

If I save the wav file and then play it it is loud.

On further investigation it appears that this is due to the fact that it is using the earpiece speaker at the top of the iPhone rather than the loudspeakers.  Is there any way I can configure the audio player to use the loudspeakers?

### LINKED ISSUE #70: IOS simple No sound,
I don't know what went wrong. I ran your sample.
And I only open the mic permission on ios, but the playback has no sound, and the playback does not stop, how should I find the problem?

https://github.com/jfversluis/Plugin.Maui.Audio/assets/35278902/c0453745-358a-4a2f-b04b-be6a8c979c72

### LINKED ISSUE #91: Playback fails on IOS devices when any audioRecorder is created
Playback of wav and mp3 files works on IOS simulators and devices, until you try to record a new audio.

As soon as this line is executed playback no longer works:

`     _audioRecorder = _audioManager.CreateRecorder();`

It doesn't matter if I go on to actually record audio, creating the recorder is enough to cause playback to fail. I can record new audio files, but they won't playback. At least not until I restart the app, at which point everything will playback including the new audio file.

MediaElement has an almost identical issue which I have already documented here:
https://github.com/CommunityToolkit/Maui/issues/1634#issuecomment-1876153420

Please see test project ReadMe for more details:
https://github.com/JamesBodi/MauiPluginAudioTestApp.git

### PR DESCRIPTION
**Pull request by bijington**
The aim of this PR is to allow developers to configure the `AVAudioSession` options when running on the Apple based platforms. It opens us up to being able to configure other platforms should we need also.

The result gives the following API changes/additions:

`AddAudio` extension method to configure the shared settings much like the options based approach in modern apps:

```csharp
var builder = MauiApp.CreateBuilder();
    builder
        .UseMauiApp<App>()
	.UseSkiaSharp()
	.AddAudio(
		playbackOptions =>
		{
#if IOS || MACCATALYST
			playbackOptions.Category = AVFoundation.AVAudioSessionCategory.Playback;
#endif
		},
		recordingOptions =>
		{
#if IOS || MACCATALYST
			recordingOptions.Category = AVFoundation.AVAudioSessionCategory.Record;
			recordingOptions.Mode = AVFoundation.AVAudioSessionMode.Default;
			recordingOptions.CategoryOptions = AVFoundation.AVAudioSessionCategoryOptions.DefaultToSpeaker;
#endif
		})
```

This fits with developers making use of dependency injection, for those that aren't there is the following:

```csharp
AudioManager.Current.SharedPlayerOptions = new
{
    Category = AVFoundation.AVAudioSessionCategory.Playback;
}
```

And then finally there is the ability to pass in an instance of the options class when creating a player or reader which will override these shared settings:

```csharp
var player = AudioManager.Current.CreatePlayer(new
    {
        Category = AVFoundation.AVAudioSessionCategory.Playback;
    });
```

Fixes #91 
Fixes #53 
Fixes #70

### CODE REVIEW

**Comment by bijington** [comment]
This is a draft for now but it would be great to hear what @jfversluis  and @borrmann think. And of course anyone else in the community :)

**Comment by borrmann** [comment]
I like that approach. It makes clear how it works on iOS. 
I think it's worth to consider one more parameter. Something like this:

```
enum SessionOptions 
{ 
  KeepSessionAlive,
  EndSession,
  EndSessionAndNotifyOthers
 }
```
KeepSessionAlive would be the current implementation. EndSession would end the audio session on FinishedPlaying and EndSessionAndNotifyOthers would end the audio session and additionally notify other audio players that listen to this event.
I don't know if there are similar methods available on Android or Windows.

**Comment by bijington** [comment]
Thanks. Is your suggestion aimed at determining whether we effectively call `AVAudioSession.SetActive();`?

**Comment by borrmann** [comment]
yes exactly, for _EndSession_ and _EndSessionAndNotifyOthers_
`AVAudioSession.SharedInstance().SetActive(false);`
would need to be called, maybe in `void OnPlayerFinishedPlaying(object? sender, AVStatusEventArgs e)` or when the audio player is disposed.
for EndSessionAndNotifyOthers, additionally, NotifyOthersOnDeactivation needs to be passed on activation.
`audioSession.SetActive(true, AVAudioSessionSetActiveOptions.NotifyOthersOnDeactivation)`

**Comment by bijington** [comment]
I haven't been able to test these changes on iOS or macOS yet but I believe this API surface should be good for a review.

File: src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.windows.cs [hunk 1]
```diff
@@ -12,6 +12,11 @@ partial class AudioRecorder : IAudioRecorder
 	public bool CanRecordAudio { get; private set; } = true;
 	public bool IsRecording => mediaCapture != null;
 
+    public AudioRecorder(AudioRecorderOptions options)
+    {
+        
+    }
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.windows.cs:18]:
```suggestion
	public AudioRecorder(AudioRecorderOptions options)
	{

	}
```

**Comment by jfversluis** [comment]
Not sure how exactly, but I guess it would be nice to surface something for this in the sample app to try out the different options as well? But for now this seems great to me!

**Comment by bijington** [comment]
> Not sure how exactly, but I guess it would be nice to surface something for this in the sample app to try out the different options as well? But for now this seems great to me!

I did update the sample to use the new approach. We could add some comments in there to cover different scenarios?

**Comment by jfversluis** [comment]
Yeah what I was thinking is more some controls to switch between the options. But that can be a follow up, if it can even be done at all? Maybe not for all options, not sure.

**Review by jfversluis — APPROVED** [review]

**Comment by bijington** [comment]
> Yeah what I was thinking is more some controls to switch between the options. But that can be a follow up, if it can even be done at all? Maybe not for all options, not sure.

Oh yes good idea! Yes we can support that

**Pull request merged**